rstudio/packrat
# Support using `GITLAB_PAT` for authentication

## Intent

Fix source downloads from private GitLab repositories by preferentially using `GITLAB_PAT` for authentication.

Fixes #673

## Approach

The GitLab downloader now checks for a `GITLAB_PAT`. If that exists, instead of creating the username/password auth, it adds `Private-Token` to the request's headers.

`GITLAB_USERNAME` and `GITLAB_PASSWORD` were removed. See [this comment](https://github.com/rstudio/packrat/pull/674#issuecomment-1169329634) for why.

## Automated Tests

There were no automated tests for GitLab download functionality before. There is one test that is skipped unless run manually.

This PR adds no tests.

## QA Notes

I validated this against a private repo on my GitLab account. Validation is tricky, requiring a Packrat lock file that restores from a GitLab repository.

To validate:

1. Download and extract this archive: [packrat-with-private-package.zip](https://github.com/rstudio/packrat/files/9003675/packrat-with-private-package.zip)
2. Get access from me to the private package hosted on GitLab that this package attempts to restore.
3. Create a [GitLab PAT](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html) and make it available in your terminal's environment as `GITLAB_PAT`.
4. Install this branch's version of Packrat.
5. Start an R session in the archive's directory that contains a `packrat` subdirectory. Run `packrat::restore()`. The environment should restore correctly.

We should also confirm:

- If the `GITLAB_PAT` environment variable is missing, an error message is displayed.
- If `GITLAB_USERNAME` and `GITLAB_PASSWORD` variables are present and `GITLAB_PAT` is absent, the same error message about `GITLAB_PAT` is displayed.

## Checklist

- [x] Added NEWS entry
